compiler/src/dotty/tools/dotc/typer/Typer.scala

```diff
@@ -1148,7 +1155,15 @@ class Typer(@constructorOnly nestingLevel: Int = 0) extends Namer
           val elsep0 = typed(tree.elsep, branchPt)(using cond1.nullableContextIf(false))
           thenp0 :: elsep0 :: Nil
         }: @unchecked
-        assignType(cpy.If(tree)(cond1, thenp1, elsep1), thenp1, elsep1)
+
+        val resType = thenp1.tpe | elsep1.tpe
+
+        val thenp2 :: elsep2 :: Nil =
+          (thenp1 :: elsep1 :: Nil) map { t =>
+            gadtAdaptBranch(t, resType)
+          }: @unchecked
+
+        cpy.If(tree)(cond1, thenp2, elsep2).withType(resType)
```

Comment:
I was considering whether we should push this logic into `assignType(If..`.  But we're already rolling our own here, by assigning the type after calling the tree copier.

Reply:
Yes, it seems that the original code is the only callsite of `assignType(If...` (not fully verified since metals often lies to me ;)). It would be good to push the logic into `assignType`.

Reply:
But it seems that pushing this logic into `TypeAssigner` will require us to also refactor the methods related to GADT casts (`gadtAdaptBranch` and `Typer.insertGadtCast`) into `TypeAssigner`. I am not sure whether it is good to put them there.

Reply:
@abgruszecki do you hold an opinion on this?

Reply:
Maybe it'd be best to quickly raise the topic during the Thursday meeting, if you think it makes sense.

Reply:
Ok! I will quickly mention it during the lab meeting.

Reply:
Yep yep, I am talking about the Thursday meeting @ 15:00.

Reply:
You should come to the meeting on Thursday and raise it! 😄 (Or I can, np)

Reply:
I'd be glad to join this lab meeting but Thursday 15:00 this week does not work well for me. :( I will appreciate knowing the results of your discussion on this topic! :D

Reply:
Martin raised a good point: seeing as this involves GADT information, that doesn't survive post typing, there's little sense to put it in a more general place like TypeAssigners: leaving it in Typer makes sense.